ZEA-2452: Add guide of how to use config (#313)
zeabur/zeabur: docs/pages/deploy/config-edit.en-US.mdx

---
title: Editing Configuration Files
ogImageTitle: Editing Configuration Files
ogImageSubtitle: Editing Configuration Files
---

import { Callout } from 'nextra-theme-docs'

# Editing Configuration Files

> When running a service, you can use a configuration file to customize some parameters of the service. Zeabur provides a configuration file editor that can specify the storage location and content of the configuration file through the editor. When the service starts, the configuration file is automatically mounted.

Reviewer: ```suggestion
When running a service, you can customize it by using a configuration file. Zeabur provides a configuration file editor that allows you to specify the path and content of your configuration. The configuration file is automatically mounted upon service startup.
```

> Here, we use [nginx](https://nginx.com/) as a front proxy for [vaultwarden](https://github.com/dani-garcia/vaultwarden) to demonstrate the process of editing ngnix custom configuration files via the configuration file editor.

Reviewer: ```suggestion
We use [NGINX](https://nginx.com/) as a front proxy for [vaultwarden](https://github.com/dani-garcia/vaultwarden) to show how to edit NGINX custom configuration files using the configuration file editor.
```

## Step One: Create a Project

Create a Project in [Zeabur Dash](https://dash.zeabur.com/).

## Step Two: Create Services

> Create Prebuilt Services in the aforementioned Project , with service names namely nginx and vaultwarden.

Reviewer: ```suggestion
Create your NGINX and Vaultwarden prebuilt services in the project you created in Step One.
```

> ![Create Ngnix Service](/deploy/config-edit/create-nginx-service.en-US.png)

Reviewer: ```suggestion
![Create NGINX Service](/deploy/config-edit/create-nginx-service.en-US.png)
```

![Create Vaultwarden Service](/deploy/config-edit/create-vultwarden-service.en-US.png)

## Step Three: Edit Nginx Service Configuration File

> 1. In the Services column, click `Settings` tab for the nginx service, click the drop-down button to find the `Configs` section, and click the `Open Config Editor` button to enter the configuration file editor.

Reviewer: ```suggestion
Click on the `Settings` tab in the Services section for your NGINX service. Then, locate the `Configs` section in the expanded tab. Next, click the `Open Config Editor` button to access the configuration file editor.
```

![Open Configuration File Editor](/deploy/config-edit/open-config-editor.en-US.png)

> 2. In the opened window, click the `Add Config file` button, to add a configuration file. In the popped-up window, input the path and name of the configuration file in the `Config Path` input box. Here, we input `/etc/nginx/nginx.conf` and click the `Save Config` button.

Reviewer: ```suggestion
2. In the editor, we click on the `Add Config file` button to add a configuration file. In the sheet that appears, enter the filename and path for your configuration in the `Config Path` text box. In this example, we will use `/etc/nginx/nginx.conf`. Finally, click the `Save Config` button.
```

![Add Configuration File](/deploy/config-edit/create-new-config.en-US.png)

<Callout>

>     The `Config Path` above requires the absolute path of the configuration file, please configure in line with the actual situation.

Reviewer: ```suggestion
    The `Config Path` above must be an absolute path. Customize it according to your requirements.
```

</Callout>

> 3. Depending on the requirements, edit the content of the configuration file. Here, we add a `server` configuration for reverse-proxying the vaultwarden service.

Reviewer: ```suggestion
3. Edit the configuration file according to your requirements. In this case, we will add a `server` configuration to enable reverse-proxying for our vaultwarden service.
```

```nginx
worker_processes  4;
error_log  stderr;
worker_rlimit_nofile 8192;

events {}

http {
  default_type application/octet-stream;
  log_format   main '$remote_addr - $remote_user [$time_local]  $status '
    '"$request" $body_bytes_sent "$http_referer" '
    '"$http_user_agent" "$http_x_forwarded_for"';
  access_log   /dev/stdout  main;
  sendfile     on;
  tcp_nopush   on;
  server_names_hash_bucket_size 128; # this seems to be required for some vhosts

  map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      "";
  }

  server {
    listen 80 default_server;
    server_name _; 

    location / {
      proxy_http_version 1.1;
      proxy_set_header Upgrade $http_upgrade;
      proxy_set_header Connection $connection_upgrade;
    
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;
    
      proxy_pass http://vaultwarden.zeabur.internal:80;
    }
  }
}
```

> In the file content above, `vaultwarden.zeabur.internal` in the `proxy_pass` block is the service name of Vaultwarden. Here, we used [Private Network](/deploy/private-networking) feature, which allows to access services through a fixed host name using an internal network.

Reviewer: ```suggestion
To allow others to access our internal Vaultwarden through our NGINX service, we use the `proxy_pass vaultwarden.zeabur.internal` directive in NGINX. This instructs NGINX to proxy all requests from `/` to Vaultwarden. We make use of the [Private Network](/deploy/private-networking) feature to access Vaultwarden with a fixed hostname within our private network.
```

> In the file above, we added a custom log format and log storage location, in addition to several custom nginx optimization parameters.

Reviewer: ```suggestion
```

I think this is unnecessary because it's derived from the default NGINX configuration of Zeabur.

> 4. Click the `Save` button to save the configuration file.

Reviewer: ```suggestion
4. Click the `Save` button to save your configuration file.
```

![Save Configuration File](/deploy/config-edit/save-config.en-US.png)

## Step Four: Restart the Service

> In the above steps, we already concluded the editing of the configuration file. Next, we need to restart the service, find the `Instructions` column, find the `Restart` button, and click to restart the service in order for the configuration file to take effect.

Reviewer: ```suggestion
We have completed the configuration of NGINX in the previous steps. Now, we need to restart the service. Go to the `Instructions` section and click the `Restart` button to apply your configuration changes.
```

![Restart Service](/deploy/config-edit/restart-service.en-US.png)

> After [binding the domain name](/deploy/domain-binding) under the nginx service, you can access the vaultwarden service through the domain name.

Reviewer: ```suggestion
Once you have [bound a domain name](/deploy/domain-binding) to your NGINX service, you can access the vaultwarden service using your domain name.
```
